drivers/mcp2515/candev_mcp2515.c

```diff
@@ -0,0 +1,570 @@
+            filter_added = 1;
+        }
+
+        /* mask existed and same mask */
+        else if (dev_mcp->masks[mailbox_index] == f.can_mask) {
+            /* find en empty space if it exist */
+            int filter_pos = 1; /* first one is already filled */
+            /* stop at the end of mailbox or an empty space found */
+            while (filter_pos < _max_filters(mailbox_index) &&
+                   dev_mcp->filter_ids[mailbox_index][filter_pos] != 0) {
+                filter_pos++;
+            }
+
+            /* an empty space is found */
+            if (filter_pos < _max_filters(mailbox_index)) {
+                /* set filter on this memory space */
+                mcp2515_set_filter(dev_mcp, MCP2515_FILTERS_MB0 * mailbox_index + filter_pos, f.can_id);
+
+                /* save filter */
+                dev_mcp->filter_ids[mailbox_index][filter_pos] = f.can_id;
+
+                /* function succeeded */
+                filter_added = 1;
+            }
+        }
+        mailbox_index++;
+    }
+
+    mcp2515_set_mode(dev_mcp, mode);
+
+    return filter_added;
+}
+
+static int _remove_filter(candev_t *dev, const struct can_filter *filter)
+{
+    DEBUG("inside _remove_filter of MCP2515\n");
+    int filter_removed = -1;
+    struct can_filter f = *filter;
+    int res = 0;
+    enum mcp2515_mode mode;
+
+    candev_mcp2515_t *dev_mcp = (candev_mcp2515_t *) dev;
+
+    if (f.can_mask == 0) {
+        return -1; /* invalid mask */
+    }
+
+    mode = mcp2515_get_mode(dev_mcp);
+    res = mcp2515_set_mode(dev_mcp, MODE_CONFIG);
+    if(res < 0) {
+        return -1;
+    }
+
+    if ((f.can_id & CAN_EFF_FLAG) == CAN_EFF_FLAG) {
+        f.can_mask &= CAN_EFF_MASK;
+    }
+    else {
+        f.can_mask &= CAN_SFF_MASK;
+    }
+
+    int mailbox_index = 0;
+    /* Browse on each mailbox to find the right filter id */
+    while (mailbox_index < MCP2515_RX_MAILBOXES && filter_removed == -1) {
+        /* same mask */
+        if (dev_mcp->masks[mailbox_index] == f.can_mask) {
+            int filter_pos = 0;
+            /* stop at the end of mailbox or filter_id found */
+            while (filter_pos < _max_filters(mailbox_index) &&
+                   dev_mcp->filter_ids[mailbox_index][filter_pos] != f.can_id) {
+                filter_pos++;
+            }
+
+            /* filter id found */
+            if (filter_pos < _max_filters(mailbox_index)) {
+                /* remove filter */
+                mcp2515_set_filter(dev_mcp, MCP2515_FILTERS_MB0 * mailbox_index + filter_pos, CAN_EFF_MASK);
+                /* save modification */
+                dev_mcp->filter_ids[mailbox_index][filter_pos] = 0;
+
+                /* check mailbox empty */
+                int nb_item = 0;
+                for (int i = 0; i < _max_filters(mailbox_index); i++) {
+                    if (dev_mcp->filter_ids[mailbox_index][i] == 0) {
+                        nb_item++;
+                    }
+                }
```

Comment:
possible optimisation (1 less variable and one evaluation of `_max_filters(...)`):

```C
int nb_item = _max_filters(mailbox_index);
while (nb_item && dev_mcp->filter_ids[mailbox_index][--nb_item] == 0);

if (nb_item == 0) { /* all 0 */ }
```